[Docs] Add an article about spec-driven development (#420)
apple/swift-openapi-generator: Sources/swift-openapi-generator/Documentation.docc/Articles/Practicing-spec-driven-API-development.md

### Publish the source of truth

Compare the workflows based on the following two properties:
1. What is the source of truth for the API? In other words, what is the representation that developers edit by hand?
2. What is the representation that the server provider publishes for the client, who consume it either through tooling or by directly reading it?

| Workflow | Source of truth | Published | Transcoding |
|----------|-----------------|-----------|-------------|
| Spec-driven | OpenAPI doc | OpenAPI doc | None required ✅ |
| Code-driven | Code | OpenAPI doc | Can be lossy and unpredictable ❌ |

Notice from the table above that the recommended spec-driven workflow publishes the same document that the developers of the server use to define the server's API (source of truth).

Publishing the source of truth is preferable to relying on transcoding from code (or other representations) to an OpenAPI document, which is often lossy and difficult to predict how a given language-specific annotation affects the OpenAPI output. It also means that any feature unsupported by the transcoder cannot be represented in the generated OpenAPI document.

> > Tip: Publish the source of truth, not a representation transcoded from the source of truth. That way, your clients can open pull requests to your OpenAPI document without having to learn how your server is implemented, nor do they need access to the source code of your server.

Author: I broke it out as you suggested, but kept the first part in a tip - visually, when rendered, it looks better to break up too many plain text paragraphs with a highlight like this.

> ### Migrate from code-driven to spec-driven development

Author: Yeah that's also be reasonable and I've considered it, but I still believe the "why" and "how" are best kept in a single article. We can also always link to sections with fragments in URLs.

#### Move over the first operation

Let's migrate the first route, `GET /foo`, and leave the other two alone for now.

First, you add the definition for the route to the OpenAPI document, so it looks something like:

```yaml
openapi: 3.1.0
info:
  title: MyService
  version: 1.0.0
paths:
  /foo:
    get:
      ... (the definition of the operation, its inputs and outputs)
```

and comment out the first of the existing route implementations in your Vapor app:

```swift
let app = Vapor.Application()

// Registers your existing routes.
// app.get("foo") { ... a, b, c ... } // <<< just comment this out, and this route will be registered below by registerHandlers, as it is now defined by your OpenAPI document.
app.post("foo") { ... a, b, c ... }
app.get("bar") { ... a, b, c ... }

struct Handler: APIProtocol {} // <<< this is where you now get a build error

let transport = VaporTransport(routesBuilder: app)
try handler.registerHandlers(on: transport, serverURL: ...)

try app.run()
```

> If you try to compile the above, you'll get a build error. Because now, the `APIProtocol` contains the requirement to implement the `getFoo` operation, but you're not implementing it, yet. Xcode will offer a Fix-it, and drop in a function stub that you just fill in:

Reviewer: ```suggestion
When you compile the example above, you'll get a build error because the `APIProtocol` contains the requirement to implement the `getFoo` operation, but it isn't yet implemented.
Xcode will offer a Fix-it, and drop in a function stub that you can fill in:
```

```swift
let app = Vapor.Application()

// Registers your existing routes.
// <<< now you can just delete the first original route, as you've moved the business logic below into the Handler type
app.post("foo") { ... a, b, c ... }
app.get("bar") { ... a, b, c ... }

struct Handler: APIProtocol {
  func getFoo(input: Operations.getFoo.Input) async throws -> Operations.getFoo.Output {
    ... b ... // <<< notice that here you just implement your business logic, but input deserialization and validation, and output serialization is handled by the generated code.
  }
}
let transport = VaporTransport(routesBuilder: app)
try handler.registerHandlers(on: transport, serverURL: ...)

try app.run()
```

Now, build run, and test!

Only one operation was moved over, but you can get confidence that it works and even deploy the service.

#### Repeat for the remaining operations

> At this point, `POST /foo` and `GET /bar` are still manually implemented, but `GET /foo` is coming from your OpenAPI document and you only had to move the business logic over.

Reviewer: ```suggestion
At this point, `POST /foo` and `GET /bar` are still manually implemented, but `GET /foo` is coming from the OpenAPI document and you only had to move the business logic over.
```

At your convenience, repeat this for the remaining two operations, until you have no manual operations (or feel free to keep some manual operations there, for example for serving static css/js files, those endpoints usually don't go into the OpenAPI document, which is meant mainly for the REST API).

#### Final state

> You should end up with something like this:

Reviewer: ```suggestion
The end result should be something like this:
```
